Source/WebCore/page/scrolling/ScrollAnchoringController.cpp

```diff
@@ -25,30 +25,96 @@
 }
 
-void ScrollAnchoringController::updateScrollPosition()
+void ScrollAnchoringController::chooseAnchorElement()
 {
-    // TODO: implement
-    // calling m_frame.scrollTo with different between m_anchorElement current position
-    // and m_lastPositionForAnchorElement
-    // should be called after suppression window (maybe Page::doAfterUpdateRendering())
+    // TODO: fully implement anchor node selection algorithm in spec
+    // (specifically excluded subtrees and priority nodes)
+    // https://drafts.csswg.org/css-scroll-anchoring/#anchor-node-selection
+
+    auto* document = m_frameView.frame().document();
+    if (!document)
+        return;
+
+    Element* candidateAnchor = nullptr;
+    auto layoutViewport = m_frameView.layoutViewportRect();
+
+    // TODO: iterate from the current scroller not document root
+    for (auto& element : descendantsOfType<Element>(document->rootNode())) {
+        if (auto renderer = element.renderBox()) {
+            if (renderer->style().overflowAnchor() == OverflowAnchor::None || &element == document->bodyOrFrameset() || is<HTMLHtmlElement>(&element))
+                continue;
+            auto boxRect = renderer->absoluteBoundingBoxRect();
+            if (layoutViewport.intersects(boxRect))
+                candidateAnchor = &element;
+            if (layoutViewport.contains(boxRect))
+                break;
+        }
+    }
+    if (!candidateAnchor)
+        return;
+    m_anchorElement = candidateAnchor;
+    m_lastOffsetForAnchorElement = computeOffset(*m_anchorElement->renderBox());
+}
+
+void ScrollAnchoringController::updateAnchorElement()
+{
+    // TODO: check owning scroller scroll position
+    if (frameView().scrollPosition().isZero() || m_isQueuedForScrollPositionUpdate)
+        return;
+
+    if (!m_anchorElement || !m_anchorElement->renderBox()) {
+        chooseAnchorElement();
+        if (!m_anchorElement)
+            return;
+    }
+    m_isQueuedForScrollPositionUpdate = true;
+    m_frameView.queueScrollableAreaForScrollAnchoringUpdate(m_frameView);
+}
+
+void ScrollAnchoringController::adjustScrollPositionForAnchoring()
+{
+    SetForScope midUpdatingScrollPositionForAnchorElement(m_midUpdatingScrollPositionForAnchorElement, true);
+    auto queued = std::exchange(m_isQueuedForScrollPositionUpdate, false);
+    if (!m_anchorElement || !queued)
+        return;
+    auto renderBox = m_anchorElement->renderBox();
+    if (!renderBox) {
+        invalidateAnchorElement();
+        return;
+    }
+    FloatSize adjustment = computeOffset(*renderBox) - m_lastOffsetForAnchorElement;
+    if (!adjustment.isZero()) {
+        auto newScrollPosition = frameView().scrollPosition() + IntPoint(adjustment.width(), adjustment.height());
+        LOG_WITH_STREAM(Scrolling, stream << "ScrollAnchoringController::adjustScrollPositionForAnchoring() for frame: " << m_frameView << " adjusting from: " << frameView().scrollPosition() << " to: " << newScrollPosition);
```

Comment:
This breaks debug builds, as `LOG_CHANNEL_PREFIXScrolling` is undeclared.

Reply:
Gotcha made https://github.com/WebKit/WebKit/pull/18062 to fix this.